Add kubebuilder tags to Condition type

damemi opened the pull request:
**What type of PR is this?**
/kind api-change


**What this PR does / why we need it**:
This picks the changes from https://github.com/kubernetes/kubernetes/pull/92519 and adds kubebuilder validation tags for CRD generation

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Kubebuilder validation tags are set on metav1.Condition for CRD generation
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```


related to https://github.com/kubernetes/enhancements/issues/1623

deads2k committed:
add validation for metav1 conditions

damemi committed:
Add kubebuilder validation tags to Condition type

damemi commented:
/hold

deads2k commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -1367,28 +1368,39 @@ type Condition struct {
 	// useful (see .node.status.conditions), the ability to deconflict is important.
 	// +required
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum=Available;Progressing;Degraded
```
this is a regex in the validation code.

damemi replied:
Yeah, I based this on the comment above the type. The validation code looks like it [uses different regexes conditionally](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go#L42) which I think is more complex than the regex marker can do alone

liggitt replied:
using an enum here is not correct... far more types than Available, Progressing, Degraded are supported

The regex validation seems to be conditional based on whether the type is qualified or not. I think that can be expressed like this:
```
(DNS1123_PATTERN/)?(QUALIFIED_NAME_PATTERN)
```

damemi replied:
@liggitt Ah, I'm not great with complex regexes so I wasn't aware if it could handle conditionals

Since dns1123 pattern is `dns1123LabelFmt + "(\\." + dns1123LabelFmt + ")*"` and qualified name fmt is `"(" + qnameCharFmt + qnameExtCharFmt + "*)?" + qnameCharFmt`, I tried substituting these into `(DNS1123_PATTERN/)?(QUALIFIED_NAME_PATTERN)` (evaluated out). This ended up being pretty complex from what I can see, not sure if there's a good way to simplify it

fedebongio commented:
/assign @liggitt @DirectXMan12

liggitt commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -1367,28 +1368,39 @@ type Condition struct {
 	// status of the condition, one of True, False, Unknown.
 	// +required
+	// +kubebuilder:validation:Required
 	Status ConditionStatus `json:"status" protobuf:"bytes,2,opt,name=status"`
```
this is an appropriate place for an enum (True, False, Unknown)

liggitt commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -1367,28 +1368,39 @@ type Condition struct {
 	// lastTransitionTime is the last time the condition transitioned from one status to another.
 	// This should be when the underlying condition changed.  If that is not known, then using the time when the API field changed is acceptable.
 	// +required
+	// +kubebuilder:validation:Required
```
is there a time format available?

damemi replied:
it looks like we can use [type: string and format: date-time](https://book.kubebuilder.io/reference/markers/crd-validation.html)

sttts commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -1346,3 +1346,64 @@ type PartialObjectMetadataList struct {
+	// The regex it matches is (dns1123SubdomainFmt/)?(qualifiedNameFmt)
+	// +required
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Pattern=^([a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*/)?(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])
```
this matches `asdfasfas.sadf.asdfas/bla$'_%`.

damemi replied:
This is what I got manually expanding the regexes currently used by the validation function (https://github.com/kubernetes/kubernetes/pull/92660#discussion_r448441739) Did I miss something or write that out wrong?

deads2k commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -1346,3 +1346,64 @@ type PartialObjectMetadataList struct {
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MaxLength=1024
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:Pattern=`^[A-Za-z_][A-Za-z0-9_]*$`
```
this got expanded `[A-Za-z]([A-Za-z0-9_,:]*[A-Za-z0-9_])?`

damemi replied:
Updated

sttts commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -1346,3 +1346,64 @@ type PartialObjectMetadataList struct {
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MaxLength=1024
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:Pattern=`^[A-Za-z_][A-Za-z0-9_]*$`
```
this doesn't match https://github.com/kubernetes/kubernetes/pull/92519/files#diff-ebee21b567c0f4e3f14894c1192dbd29R251

damemi replied:
Updated (https://github.com/kubernetes/kubernetes/pull/92660#discussion_r449065755)

damemi commented:
@deads2k @sttts think I addressed the comments here. There was some offline talk about trying to come up with a regex to match specifically non-zero rfc3339, is that still something we need or is a date-time format validation sufficient? (https://github.com/kubernetes/kubernetes/pull/92660#discussion_r448512878)

damemi commented:
/retest

deads2k commented:
Can you rebase this on master please?  You've got an old version of the validation PR

damemi commented:
> Can you rebase this on master please? You've got an old version of the validation PR

Ah I did rebase it on master, I needed to rebase it on your branch from the validation PR. Updated and squashed

deads2k commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -1369,29 +1369,44 @@ type Condition struct {
+	// The regex it matches is (dns1123SubdomainFmt/)?(qualifiedNameFmt)
 	// +required
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Pattern=`^([a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*/)?(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])$`
```
you sure you need the `\\`?  Can you make a proof PR we can play with?

damemi replied:
fixed

deads2k commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -1369,29 +1369,44 @@ type Condition struct {
 	// useful (see .node.status.conditions), the ability to deconflict is important.
+	// The regex it matches is (dns1123SubdomainFmt/)?(qualifiedNameFmt)
 	// +required
+	// +kubebuilder:validation:Required
```
it's not perfect, but the max length is 253+63=316

damemi replied:
added

deads2k commented:
also `make update`.  your generated code is out of date

damemi commented:
@deads2k updated my commit and rebased on yours again

deads2k commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -1369,29 +1369,45 @@ type Condition struct {
 	// message is a human readable message indicating details about the transition.
 	// This may be an empty string.
 	// +required
+	// +kubebuilder:validation:Required
```
we allow the message to be empty.  See two lines up.  The required bit is about serialization back to clients.

liggitt replied:
allowed to be empty, but the field must be specified or openapi validation will complain (`"message":""`)

deads2k replied:
> This looks correct to me. The value can be an empty string, but the field must be specified or openapi validation will complain (`"message":""`)

Found it in my live test.  The CR validation doesn't give slice numbers, so I was looking at the wrong item.

deads2k commented:
/lgtm

deads2k commented:
/retest

smarterclayton commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/generated.proto:
```diff
@@ -177,11 +187,17 @@ message Condition {
+  // +kubebuilder:validation:Required
+  // +kubebuilder:validation:MaxLength=1024
+  // +kubebuilder:validation:MinLength=1
+  // +kubebuilder:validation:Pattern=`^[A-Za-z]([A-Za-z0-9_,:]*[A-Za-z0-9_])?$`
```
Allowing commas.  Not a fan....

smarterclayton replied:
Same for colon.  Hrm... gonna have to think about this.

deads2k replied:
> Allowing commas. Not a fan....

Matches the reality of the world I'm afraid.  We have people interested in adopting now that have union reasons with delimiters

smarterclayton replied:
I guess since we said "machine readable" this is the world we live in.

smarterclayton commented:
/approve

deads2k commented:
/retest

deads2k commented:
validation is finalized, removing hold.

damemi commented:
/label api-review
